### PR TITLE
[Core][Distributed] refactor custom allreduce to support multiple tp groups

### PR DESCRIPTION
**Pull request by youkaichao**
Previously custom allreduce is attached to a module, and only bound to the world group.

With this PR, it is bound correctly to the tp group.

- [ ] remove import inside function (need to remove default group None option)

### CODE REVIEW

**Comment by WoosukKwon** [comment]
@hanzhi713 Could you please also take a look if you have time?

File: vllm/distributed/parallel_state.py [hunk 1]
```diff
@@ -162,12 +176,31 @@ def initialize_model_parallel(
...
 
+    # A small all_reduce for warmup.
+    # this is needed because device communicators might be created lazily
+    # (e.g. NCCL). This will ensure that the communicator is initialized
+    # before any communication happens, so that this group can be used for
+    # graph capture immediately.
+    data = torch.zeros(1)
+    if torch.cuda.is_available():
+        data = data.to(device=f"cuda:{_LOCAL_RANK}")
+    torch.distributed.all_reduce(data, group=_TP_DEVICE_GROUP)
```
Comment (youkaichao) [inline vllm/distributed/parallel_state.py:187]: add torch.cuda.sync after every warmup

Comment (youkaichao) [reply vllm/distributed/parallel_state.py:187]: fixed in [8fd22fd](https://github.com/vllm-project/vllm/pull/4754/commits/8fd22fd2d29b52038c6c41f23b2d70894d051fc2).

File: tests/distributed/test_custom_all_reduce.py [hunk 2]
```diff
@@ -41,44 +50,54 @@ def graph_allreduce(world_size, rank, distributed_init_port):
...
 
 
 @ray.remote(num_gpus=1, max_calls=1)
-def eager_allreduce(world_size, rank, distributed_init_port):
+def eager_allreduce(tp_size, pp_size, rank, distributed_init_port):
     del os.environ["CUDA_VISIBLE_DEVICES"]
     device = torch.device(f"cuda:{rank}")
     torch.cuda.set_device(device)
-    init_test_distributed_environment(1, world_size, rank,
+    init_test_distributed_environment(pp_size, tp_size, rank,
```
Comment (WoosukKwon) [inline tests/distributed/test_custom_all_reduce.py:70]: nit: can we make the order of `tp_size` and `pp_size` consistent?

File: vllm/distributed/parallel_state.py [hunk 3]
```diff
@@ -162,12 +176,33 @@ def initialize_model_parallel(
...
+
     from vllm.distributed.device_communicators.pynccl import PyNcclCommunicator
     _TP_PYNCCL_COMMUNICATOR = PyNcclCommunicator(
         group=_TP_CPU_GROUP,
         device=_LOCAL_RANK,
     )
 
+    # Initialize a custom fast all-reduce implementation.
+    if _ENABLE_CUSTOM_ALL_REDUCE:
+        from vllm.distributed.device_communicators.custom_all_reduce import (
```
Comment (WoosukKwon) [inline vllm/distributed/parallel_state.py:199]: Why do we need lazy import here?

File: vllm/distributed/communication_op.py [hunk 4]
```diff
@@ -24,10 +25,18 @@ def graph_capture_mode():
...
+    # We always prioritize using custom all-reduce kernel but fall back
+    # to PyTorch or pynccl if it is disabled or not supported.
     pynccl_comm = get_tp_pynccl_communicator()
+    ca_comm = get_tp_ca_communicator()
     assert pynccl_comm is not None
+    assert ca_comm is not None
     with pynccl_comm.change_state(enable=True,
-                                  stream=torch.cuda.current_stream()):
+                                  stream=torch.cuda.current_stream()), \
+                                  ca_comm.capture():
```
Comment (WoosukKwon) [inline vllm/distributed/communication_op.py:39]: nit:
```suggestion
    with (pynccl_comm.change_state(enable=True,
                                  stream=torch.cuda.current_stream()), 
                                  ca_comm.capture()):
```

File: tests/distributed/test_custom_all_reduce.py [hunk 5]
```diff
@@ -18,17 +20,24 @@
...
+    group = get_tensor_model_parallel_group()
+    # we use the first group to communicate once
+    # and the second group to communicate twice
+    # and so on
+    # this is used to demonstrate that each group can
+    # communicate independently
+    num_communication = rank // tp_size + 1
+
     for sz in test_sizes:
         for dtype in [torch.float32, torch.float16, torch.bfloat16]:
```
Comment (WoosukKwon) [inline tests/distributed/test_custom_all_reduce.py:39]: nit: Maybe we can use `dtype` as an input parameter so that the dtypes can be tested separately?

File: vllm/distributed/device_communicators/custom_all_reduce.py [hunk 6]
```diff
@@ -188,30 +74,117 @@ def _can_p2p(rank: int, world_size: int) -> bool:
...
+        self._IS_CAPTURING = False
+        self.disabled = True
+
+        if custom_ar is None:
+            # disable because of missing custom allreduce library
+            # e.g. in a non-cuda environment
+            return
+
+        from vllm.distributed.parallel_state import (
+            get_tensor_model_parallel_cpu_group)
```
Comment (WoosukKwon) [inline vllm/distributed/device_communicators/custom_all_reduce.py:103]: Could you say again why we need this lazy import?

File: vllm/distributed/device_communicators/custom_all_reduce.py [hunk 7]
```diff
@@ -269,8 +252,31 @@ def all_reduce_unreg(self, inp: torch.Tensor, out: torch.Tensor = None):
...
+            # be small(<=1% of overall latency) compared to the performance
+            # gains of using custom kernels
+            if self.should_custom_ar(input):
+                return self.all_reduce_unreg(input)
+
+        return None
+
     def close(self):
-        if self._ptr:
+        if not self.disabled and self._ptr:
```
Comment (WoosukKwon) [inline vllm/distributed/device_communicators/custom_all_reduce.py:279]:
```suggestion
        if custom_ar is not None and self._ptr:
```

File: vllm/distributed/device_communicators/custom_all_reduce.py [hunk 8]
```diff
@@ -188,30 +74,117 @@ def _can_p2p(rank: int, world_size: int) -> bool:
...
+        if cuda_visible_devices:
+            device_ids = list(map(int, cuda_visible_devices.split(",")))
+        else:
+            device_ids = list(range(torch.cuda.device_count()))
+
+        physical_device_id = device_ids[device.index]
+        tensor = torch.tensor([physical_device_id], dtype=torch.int)
+        gather_list = [
+            torch.tensor([0], dtype=torch.int) for _ in range(world_size)
+        ]
```
Comment (WoosukKwon) [inline vllm/distributed/device_communicators/custom_all_reduce.py:143]: nit: Add `device="cpu"` in `torch.tensor` for clarity?

**Review by WoosukKwon — COMMENTED** [review]
@youkaichao Thanks for submitting the PR and many thanks for walking through the PR offline. Please check out my comments, most of which are style issues.

File: tests/distributed/test_comm_ops.py [hunk 9]
```diff
@@ -16,15 +16,15 @@
 
 
 @ray.remote(num_gpus=1, max_calls=1)
-def all_reduce_test_worker(tensor_parallel_size: int, rank: int,
+def all_reduce_test_worker(tensor_parallel_size: int, pp_size: int, rank: int,
```
[the remark this answers is not in the document]
Comment (youkaichao) [reply tests/distributed/test_comm_ops.py]: fixed in [129eada](https://github.com/vllm-project/vllm/pull/4754/commits/129eada32bed5b7acf9dca648d8b375ce79ecf29)

**Comment by bingfengyiren** [comment]
您的邮件已收到，我会尽快回复！

File: tests/distributed/test_custom_all_reduce.py (hunk 2 above)
Comment (youkaichao) [reply tests/distributed/test_custom_all_reduce.py]: fixed in [606c732](https://github.com/vllm-project/vllm/pull/4754/commits/606c7329793a0211255b77e2717e068003b893d3)

File: vllm/distributed/communication_op.py (hunk 4 above)
Comment (youkaichao) [reply vllm/distributed/communication_op.py]: fixed in [105d14c](https://github.com/vllm-project/vllm/pull/4754/commits/105d14cfe97924f762771d02f98299b07e144cf7).

File: vllm/distributed/device_communicators/custom_all_reduce.py (hunk 8 above)
Comment (youkaichao) [reply vllm/distributed/device_communicators/custom_all_reduce.py]: fixed in [0b9950f](https://github.com/vllm-project/vllm/pull/4754/commits/0b9950f3a1f1c61a1e5f1a9882766ac83125c8a0)

File: vllm/distributed/parallel_state.py (hunk 3 above)
Comment (youkaichao) [reply vllm/distributed/parallel_state.py:199]: The circular import:

`vllm/distributed/__init__.py` imports `parallel_state` and `communication_op`. If `parallel_state` imports from `vllm.distributed.device_communicators.custom_all_reduce` in the top level, then this is a circular import because `custom_all_reduce` imports `get_tensor_model_parallel_cpu_group` from `parallel_state`.

Therefore, either `parallel_state` or `custom_all_reduce` has to use lazy import to break the circular import.

I use lazy import in `parallel_state`, to be consistent with how we import `pynccl`.

File: vllm/distributed/device_communicators/custom_all_reduce.py (hunk 6 above)
Comment (youkaichao) [reply vllm/distributed/device_communicators/custom_all_reduce.py]: fixed in [f286b73](https://github.com/vllm-project/vllm/pull/4754/commits/f286b739e46a54ab4b44188a95312d4797989608)

File: vllm/distributed/device_communicators/custom_all_reduce.py (hunk 7 above)
Comment (youkaichao) [reply vllm/distributed/device_communicators/custom_all_reduce.py:305]: fixed in [dd71361](https://github.com/vllm-project/vllm/pull/4754/commits/dd71361dead3d295aa2c63a2c029a3d57c047fbc).  `custom_ar is None` implies `self.disabled == True`

File: tests/distributed/test_custom_all_reduce.py (hunk 5 above)
Comment (youkaichao) [reply tests/distributed/test_custom_all_reduce.py:39]: Quite difficult. The input of `test_custom_allreduce`, is coupled with the input of `multi_process_tensor_parallel`, which is used elsewhere. In other words, we cannot modify the input parameter inside just `tests/distributed/test_custom_all_reduce.py` 🤣

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for addressing my comments!

**Comment by hanzhi713** [comment]
LGTM
